doc/making-a-visualizer.md

```diff
@@ -289,3 +250,72 @@ In a documentation block just before the first event handling method, put
 You can organize the control documentation comments in whatever way makes the
 most sense. If possible, order the event handling methods so that most
 controls can be documented near the method that implements them.
+
+### Abstract visualizers
+
+Now that we've seen how to extend the [`P5Visualizer`](#p5-visualizers) base
+class, let's take a peek at how the base class works internally. This section
+will be most useful to you if you want to write a new base class, or to build
+a visualizer so different from anything else that it shouldn't have a base
+class.
+
+Behind the scenes, a visualizer base class is an implementation of the
+visualizer interface (`VisualizerInterface`). To support parameters, the base
+class also has to implement the parameterizable object interface
+(`ParamableInterface`). These interfaces are defined in
+`VisualizerInterface.ts` and `Paramable.ts`, respectively. To write a new base
+class, or to build a visualizer without one, you'll have to implement these
+interfaces yourself. That means your visualizer class has to include the
+following data and methods, and they have to behave in the way the engine
+expects.
+
+#### The visualizer interface
+
+1. `visualization()`: Returns a string saying what type of visualizer this is.
+   Typically, each base class chooses its own fixed return value, which all
+   its descendants inherit.
+2. `view(seq)`: Load the given sequence into the visualizer, where the drawing
+   operations in later function calls will be able to access them. This method
+   should not do any drawing.
+3. `inhabit(element)`: Insert a view of the the visualizer into the given
+   `HTMLElement`. This element is typically a `div` whose size is already set
+   up to comprise the available space for visualization. The `inhabit()`
+   method should not do any drawing.
+4. `show()`: Start or resume display of the visualization.
+5. `stop()`: Pause display of the visualization. Don't erase any visualization
+   produced so far or otherwise clean up the visualizer.
+6. `depart()`: Throw out the visualization, release its resources, remove its
+   injected DOM elements, and do any other required cleanup. After this call,
+   the visualizer must support `inhabit()` being called again, perhaps with a
+   different div, to re-initialize the visualization.
+
+#### The paramable interface
+
+1. `name`
+2. `description`
+3. `isValid`: A boolean that is used to determine if the visualizer is ready
+   to draw. Generally this will be set automatically based on what you return
+   from the `checkParameters` method (see below).
+4. `params`: The engine expects all visualizers to have parameters that can be
+   set by the user, though these parameters can be empty. This `params`
```

Comment:
same comment about not mentioning visualizers here. Please reword. Remember I will be filing an issue to either link to this documentation within, or extract it from, Paramable.ts where it belongs.

Reply:
I've removed all the other references t visualizers. In the process, I realized that the description of `validate()` had been confusing me, so I rewrote it.

It might be nice to rename of `validate()` at some point, because the current name made me surprised that `validate()` is supposed to call `assignParameters()` if appropriate. I usually think of validation as something you do before loading parameters—not something you do to cause parameters to be loaded.

Reply:
You are certainly welcome to file an issue/PR about these method names and division of labor. It is true there was behavior creep -- we were trying to be prepared for when upon entry/change of a single parameter, it would immediately validate _and_ immediately be transmitted to the visualizer so it could take immediate effect. So I think we definitely need some method that has that effect, whatever it be called. Resolving for now.